### PR TITLE
zephyr: Roaming support

### PR DESCRIPTION
**Pull request by nxf58150**
Added roaming support in wpa_supplicant for Zephyr.

### CODE REVIEW

**Comment by nxf58150** [comment]
The 11kvr roaming RFC link: https://github.com/zephyrproject-rtos/zephyr/issues/76307

File: src/drivers/driver_zephyr.c [hunk 1]
```diff
@@ -530,11 +530,38 @@ void wpa_drv_zep_event_proc_auth_resp(struct zep_drv_if_ctx *if_ctx,
 }
 
 
+static void wpa_drv_zep_free_pairwise_key_params(struct wpa_driver_set_key_params *params)
+{
+	if (params->ifname) {
```
[the remark this answers is not in the document]
Comment (nxf58150) [reply src/drivers/driver_zephyr.c]: Removed all checks. Thanks!

**Review by jukkar — APPROVED** [review]

**Comment by jukkar** [comment]
ping @krish2718, please take a look

**Review by krish2718 — COMMENTED** [review]
Can you please split the 2nd commit in to two? The commit log as bulleted list, a good indication of the need to split the commit.

**Comment by nxf58150** [comment]
> Can you please split the 2nd commit in to two? The commit log as bulleted list, a good indication of the need to split the commit.

Done. Please help review it again. Thanks!

**Comment by jukkar** [comment]
I think this roaming PR could go next, @krish2718 please take a look of this and also the zephyr counter part.

File: src/drivers/driver_zephyr.c [hunk 2]
```diff
@@ -1720,6 +1817,29 @@ static int _wpa_drv_zep_set_key(void *priv,
 static int wpa_drv_zep_set_key(void* priv,
 			       struct wpa_driver_set_key_params *params)
 {
+	struct zep_drv_if_ctx *if_ctx = NULL;
+	enum key_flag key_flag        = params->key_flag;
+
+	if_ctx = priv;
+
+	if (if_ctx->ft_roaming && (key_flag & KEY_FLAG_PAIRWISE)) {
+		return wpa_drv_zep_save_pairwise_key_params(priv, params);
```
Comment (krish2718) [inline src/drivers/driver_zephyr.c:1826]: The issue itself looks generic enough but this fix is a hack, without using extended keys (storing both keys in parallel) the packet loss is common either in re-keying or roaming, do we know how the nl80211 handles this scenario? I can also check that and see if we can come up with a proper solution than storing and installing keys at a later point.

Comment (nxf58150) [reply src/drivers/driver_zephyr.c]: I removed the fix in zephyr driver and added it to our wifi driver. If other vendor uses zephyr also has this issue, we can discuss more generic fix in zephyr driver in the future. Thanks!

**Comment by MaochenWang1** [comment]
hi @krish2718  please review this, we need this PR to make the upstream NXP wifi example build pass.

File: src/drivers/driver_zephyr.c [hunk 3]
```diff
@@ -1812,7 +1832,11 @@ static int wpa_drv_zep_set_supp_port(void *priv,
 
 #ifdef CONFIG_NET_DHCPV4
 	if (authorized) {
-		net_dhcpv4_restart(iface);
+		if (if_ctx->ft_roaming == false) {
+			net_dhcpv4_restart(iface);
+		} else {
+			if_ctx->ft_roaming = false;
+		}
```
Comment (krish2718) [inline src/drivers/driver_zephyr.c:1839]: This only works for L2 roaming, should this be a Kconfig option? or an alternative solution?

Comment (nxf58150) [reply src/drivers/driver_zephyr.c:1839]: What do you mean by Kconfig option?

Comment (krish2718) [reply src/drivers/driver_zephyr.c:1839]: Something like below? Also please extend the commit log to explain

```
config WIFI_NM_WPA_SUPPLICANT_SKIP_DHCP_ON_ROAMING
   bool "Do not restart DHCP when roaming e.g., L2 roaming"
   help
     For L2 roaming, the IP network doesn't change, so, DHCP can be skipped.
     For L3 roaming this needs to be enabled.
```

Comment (nxf58150) [reply src/drivers/driver_zephyr.c:1839]: Could you please help clarify L2 roaming and L3 roaming? Thanks!

Comment (krish2718) [reply src/drivers/driver_zephyr.c:1839]: L2 romaing, romaing within same subnet, L3 roaming is roaming across subnets, some info https://learningnetwork.cisco.com/s/question/0D56e0000CjUe4lCQC/what-is-layer2-and-layer3-roaming . The point is this decision of whether to perform DHCP renewal or not should be user configurable as each scenario might need a different behaviour.

Comment (nxf58150) [reply src/drivers/driver_zephyr.c:1839]: Ok, got it. I will update the PR later. Thanks!

Comment (nxf58150) [reply src/drivers/driver_zephyr.c:1840]: Hi @krish2718 , I added new kconfig WIFI_NM_WPA_SUPPLICANT_SKIP_DHCP_ON_ROAMING as suggested and wrapped the source code in this new config. Please help review it. Thanks!

**Pull request merged**